### PR TITLE
[MODINREACH-162] INN-Reach Circulation Flow: React to Loan Record Changes for INN-Reach Transactions - Local Hold: Check Out to Local Patron

### PR DESCRIPTION
**Pull request by aleksandr2110**
## Purpose
https://issues.folio.org/browse/MODINREACH-162

## Approach
When a item associated with a Local Hold transaction in INN-Reach is checked out to the requesting patron via the FOLIO mod-circulation APIs (thereby creating a new loan record in loan-storage (mod-circulation-storage), mod-inn-reach needs to associate the loan record with the INN-Reach transaction (set the loanId on the transaction record) and issue the appropriate terminating transaction request to the central server.
1. mod-inn-reach should consume loan-storage domain events via Kafka ([CIRCSTORE-300](https://issues.folio.org/browse/CIRCSTORE-300))
2. If the associated INN-Reach transaction is a LOCAL_HOLD, issue an local check out message to the central server (terminating transaction, D2IR API Reference, pg. 57) and update the transaction status to LOCAL_CHECKOUT.
## Acceptance criteria:
New loans of INN-Reach items are reliably associated with the appropriate INN-Reach transaction when created
Local items checked out to local patrons for which there is a LOCAL_HOLD transaction in INN-Reach trigger the appropriate message to the central server and update the status of the INN-Reach transaction accordingly (LOCAL_CHECKOUT)

## Pre-Merge Checklist:
Before merging this PR, please go through the following list and take appropriate actions.

- Does this PR meet or exceed the expected quality standards?
  - [x] Code coverage on new code is 80% or greater
  - [x] Duplications on new code is 3% or less
  - [x] There are no major code smells or security issues
- Does this introduce breaking changes?
  - [ ] Were any API paths or methods changed, added or removed?
  - [ ] Were there any schema changes?
  - [ ] Did any of the interface versions change?
  - [ ] Were permissions changed, added, or removed?
  - [ ] Are there new interface dependencies?
  - [ ] There are no breaking changes in this PR.
  
If there are breaking changes, please **STOP** and consider the following:

- What other modules will these changes impact?
- Do JIRAs exist to update the impacted modules?
  - [ ] If not, please create them
  - [ ] Do they contain the appropriate level of detail?  Which endpoints/schemas changed, etc.
  - [ ] Do they have all they appropriate links to blocked/related issues?
- Are the JIRAs under active development?  
  - [ ] If not, contact the project's PO and make sure they're aware of the urgency.
- Do PRs exist for these changes?
  - [ ] If so, have they been approved?

Ideally all of the PRs involved in breaking changes would be merged in the same day to avoid breaking the folio-testing environment.  Communication is paramount if that is to be achieved, especially as the number of intermodule and inter-team dependencies increase.  

While it's helpful for reviewers to help identify potential problems, ensuring that it's safe to merge is ultimately the responsibility of the PR assignee.

### CODE REVIEW

File: src/test/java/org/folio/innreach/domain/listener/KafkaCirculationEventListenerApiTest.java [hunk 1]
```diff
@@ -72,18 +73,22 @@
 @SqlMergeMode(MERGE)
 class KafkaCirculationEventListenerApiTest extends BaseKafkaApiTest {
   private static final UUID LOAN_ID = UUID.randomUUID();
+  private static final UUID PRE_POPULATED_LOAN_ID = UUID.fromString("7b43b4bc-3a57-4506-815a-78b01c38a2a1");
```
Comment (oliinyko) [inline src/test/java/org/folio/innreach/domain/listener/KafkaCirculationEventListenerApiTest.java]: PRE_POPULATED_**LOCAL**_...

File: src/test/java/org/folio/innreach/domain/listener/KafkaCirculationEventListenerApiTest.java [hunk 2]
```diff
@@ -72,18 +73,22 @@
...
   private static final UUID REQUESTER_ID = UUID.randomUUID();
   private static final UUID ITEM_ID = UUID.fromString("8a326225-6530-41cc-9399-a61987bfab3c");
+  private static final UUID PRE_POPULATED_ITEM_ID = UUID.fromString("c633da85-8112-4453-af9c-c250e417179d");
```
Comment (oliinyko) [inline src/test/java/org/folio/innreach/domain/listener/KafkaCirculationEventListenerApiTest.java]: PRE_POPULATED_**LOCAL**_...

File: src/main/java/org/folio/innreach/domain/service/impl/InnReachTransactionActionServiceImpl.java [hunk 3]
```diff
@@ -225,6 +230,25 @@ public void handleCheckInCreation(CheckInDTO checkIn) {
...
+  }
+
+  private void associateNewLoanWithLocalTransaction(StorageLoanDTO loan, InnReachTransaction transaction) {
+    if (transaction.getState() == LOCAL_HOLD) {
+      log.info("Associating a new loan {} with local transaction {}", loan.getId(), transaction.getId());
+      var hold = transaction.getHold();
+      hold.setFolioLoanId(loan.getId());
+      transaction.setState(LOCAL_CHECKOUT);
+      var inventoryItemDTO = itemService.find(loan.getItemId())
+        .orElseThrow(() -> new IllegalArgumentException("Item is not found by id " + loan.getItemId()));
```
Comment (oliinyko) [inline src/main/java/org/folio/innreach/domain/service/impl/InnReachTransactionActionServiceImpl.java]: Duplicate code. Use `fetchItemById`

File: src/main/java/org/folio/innreach/domain/service/impl/InnReachTransactionActionServiceImpl.java [hunk 4]
```diff
@@ -225,6 +230,25 @@ public void handleCheckInCreation(CheckInDTO checkIn) {
...
 
+  private void associateNewLoanWithPatronTransaction(StorageLoanDTO loan, InnReachTransaction transaction) {
+    log.info("Associating a new loan {} with patron transaction {}", loan.getId(), transaction.getId());
+    var hold = transaction.getHold();
+    hold.setFolioLoanId(loan.getId());
+    hold.setDueDateTime(toEpochSec(loan.getDueDate()));
+  }
+
+  private void associateNewLoanWithLocalTransaction(StorageLoanDTO loan, InnReachTransaction transaction) {
+    if (transaction.getState() == LOCAL_HOLD) {
```
Comment (oliinyko) [inline src/main/java/org/folio/innreach/domain/service/impl/InnReachTransactionActionServiceImpl.java]: Not sure if the requirements about `LOCAL_HOLD` refers to transaction state or type. Clarify with PO if there is need to check the state

> If the associated INN-Reach transaction is a LOCAL_HOLD

File: src/test/java/org/folio/innreach/domain/listener/KafkaCirculationEventListenerApiTest.java (hunk 2 above)
Comment (aleksandr2110) [reply src/test/java/org/folio/innreach/domain/listener/KafkaCirculationEventListenerApiTest.java]: Fixed

File: src/test/java/org/folio/innreach/domain/listener/KafkaCirculationEventListenerApiTest.java (hunk 1 above)
Comment (aleksandr2110) [reply src/test/java/org/folio/innreach/domain/listener/KafkaCirculationEventListenerApiTest.java]: Fixed

File: src/main/java/org/folio/innreach/domain/service/impl/InnReachTransactionActionServiceImpl.java (hunk 3 above)
Comment (aleksandr2110) [reply src/main/java/org/folio/innreach/domain/service/impl/InnReachTransactionActionServiceImpl.java]: Fixed

File: src/main/java/org/folio/innreach/domain/service/impl/InnReachTransactionActionServiceImpl.java (hunk 4 above)
Comment (aleksandr2110) [reply src/main/java/org/folio/innreach/domain/service/impl/InnReachTransactionActionServiceImpl.java]: Fixed

**Review by oliinyko — APPROVED** [review]

**Pull request merged**